Improve how we record and handle entity_history

lutter opened the pull request:
The main focus of this change is to stop recording entity_history for the
'subgraphs' subgraph, and to avoid recording history for changes that
actually do not change the entity. The history we recorded for these made
up about 75% of the storage needed in the hosted offering, and we never
need the data thus recorded.

This change also simplifies how the revert_block stored procedure works by
basing it off a single history query, and removes unused input parameters
for it.

The change also gets rid of a few stored procedures that were previously
used, but are not any longer.

More background on these changes can be found at
https://github.com/graphprotocol/graph-node/issues/807

lutter commented:
This PR addresses a few of the bigger points from issue #807. 

Note that the migration that comes with this PR runs for a very long time (~10 minutes on my machine, caused by the delete statements at the end of the migration). It should not affect normal operation of the service as Postgres will only lock the rows that we are deleting, and we never touch them, but it will have an impact on CPU/memory.

To make this even scarier, we will need to issue a `VACUUM (FULL, ANALYZE) entity_history` and a `VACUUM (FULL, ANALYZE) event_meta_data` to reclaim the space from all the deleted rows at some point. But that will make the service inoperable as it will prevent any read or write activity to these tables while it is running. We could also do just a 'normal' vacuum, not a full one, which will still allow normal operation, and only prohibit schema changes while it is running. It will not shrink the storage on disk, but should otherwise be just as good.

Jannis commented on store/postgres/migrations/2019-03-07-171355_update_history_storage/up.sql:
```diff
@@ -0,0 +1,178 @@
+-- Create a view to get entity_history by source ordered by event_id desc
+-- We use this both in revert_block and in the Rust code in
+-- store_events.get_revert_event
+create or replace view entity_history_with_source as
```
Nit: Should the SQL keywords such as `CREATE OR REPLACE VIEW` be all-uppercase here, to be consistent with the rest of the file?

lutter replied:
Heh. I keep going back and forth on this, and find the all-caps style incredibly shouty (it's not the 80's anymore, we do have terminals with lower and uppercase) but I'll change that to match the rest of the file.

lutter replied:
I solved that by downcasing everything in the `up` migration. The `down` migration is just copy pasta of stuff from previous migrations.

Jannis commented on store/postgres/migrations/2019-03-07-171355_update_history_storage/up.sql:
```diff
@@ -0,0 +1,178 @@
+
+    new_event_id := null;
+
+    IF event_id IS NULL THEN
```
I know this was here before, but now that I look at it, I don't understand what this (and the `SELECT` query just above) does. Can we add a comment?

lutter replied:
Yeah, a comment is in order. It's making sure that we have a `meta_data_event` for the current transaction, but that we only create one per transaction (so that all `entity_history` entries that get created in the txn are tied to the same event)

Jannis commented on store/postgres/migrations/2019-03-07-171355_update_history_storage/up.sql:
```diff
@@ -0,0 +1,178 @@
+    FROM entity_history_with_source h
+    WHERE h.source = block_to_revert_hash
+      AND h.subgraph = subgraph_id
+    LOOP
```
I was going to say "these were ordered before" and then I realized `entity_history_with_source` _is_ pre-ordered. It would be nice to have a comment here that expresses that _or_ to rename `entity_history_with_source to `ordered_entity_history_with_source` to make this obvious.

lutter replied:
Ok. I'll add a comment, both here and where the view is defined.

lutter commented:
There's one more thing that I need to address: since we are deleting a ton of data, we will likely run into a scenario where the WAL log becomes way too big while the `delete` is running. I think I need to batch the delete (say 1,000,000 rows at a time) and commit those explicitly to avoid WAL issues.

lutter commented:
I've decided to change strategy a little bit: rather than bundle everything into one PR (and one migration), I am making this PR only about not recording history that we shouldn't be recording in the first place, and will open another PR for the migration that will remove the needlessly recorded data. These two operations carry hugely different risks.

As it is now, this PR is a fairly routine code change, and can be roled out without putting much stress on the database. I believe I have addressed all review comments so far, but would hugely appreciate another thorough review of the changes.

lutter commented:
Rebased to latest master.
